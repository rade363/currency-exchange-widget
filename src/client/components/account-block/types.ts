import { Currency } from '../../utils/create-currency/types';

export type Account = {
  balance: number;
  currency: Currency;
};

export type OwnProps = {
  type: 'From' | 'To';
  currentAccount: Account;
  otherAccount: Account;
};
