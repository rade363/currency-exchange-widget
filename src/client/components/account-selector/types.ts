import { Account } from '../../hooks/useExchange/types';
import { AccountType } from '../account-block/types';
import { UserAccount } from '../app/types';

export type OwnProps = {
  type: AccountType;
  accountToReplace: Account;
  accounts: UserAccount[];
  closeSelector: () => void;
};
