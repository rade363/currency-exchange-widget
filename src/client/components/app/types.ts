import { CurrencyName } from '../../utils/create-currency/types';

export type User = {
  accounts: UserAccount[];
};

export type UserAccount = {
  balance: number;
  name: CurrencyName;
};
