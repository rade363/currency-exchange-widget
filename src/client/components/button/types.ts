import { ButtonHTMLAttributes } from 'react';

export type OwnProps = {
  active: boolean;
  styleType: 'primary' | 'danger';
} & ButtonHTMLAttributes<HTMLButtonElement>;
