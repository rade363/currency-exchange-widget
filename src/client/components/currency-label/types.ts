import { ButtonHTMLAttributes } from 'react';
import { Currency } from '../../utils/create-currency/types';

export type OwnProps = {
  currency: Currency;
  showArrow?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;
