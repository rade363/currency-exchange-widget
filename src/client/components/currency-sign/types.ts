import { HTMLAttributes } from 'react';

export type OwnProps = {
  sign: string;
} & HTMLAttributes<HTMLDivElement>;
