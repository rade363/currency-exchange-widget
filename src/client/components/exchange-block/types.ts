import { Dispatch, SetStateAction } from 'react';
import { RatesTable } from '../../utils/calculate-rates/types';
import { UserAccount } from '../app/types';

export type SetAccountsCallback = Dispatch<SetStateAction<UserAccount[]>>;

export type OwnProps = {
  rates: RatesTable;
  accounts: UserAccount[];
  setAccounts: SetAccountsCallback;
};
