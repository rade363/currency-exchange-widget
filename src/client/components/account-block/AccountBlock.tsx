import React, { useCallback, ChangeEvent } from 'react';
import CurrencyLabel from '../currency-label';
import { OwnProps } from './types';
import { Nullable } from '../../../types';
import './account-block.scss';

function setStatus(error: Nullable<string>, result: Nullable<string>): string {
  if (error) {
    return error;
  }

  return result === null ? '' : `Result: ${result}`;
}

function AccountBlock({ type, currentAccount, otherAccount, onInputChange, handleSelectorClick }: OwnProps) {
  const { balance, change, result, error, currency } = currentAccount;

  const handleChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    onInputChange(event.currentTarget.value);
  }, [onInputChange]);

  const handleBlur = useCallback(() => {
    if (change.indexOf('.') < 0) {
      const newNumber = Number(change).toFixed(2);
      onInputChange(newNumber);
      return;
    }

    const enteredValueArr = change.split('.');
    const [, decimal] = enteredValueArr;

    if (decimal.length === 2) {
      return;
    }

    onInputChange(Number(change).toFixed(2));
  }, [change, onInputChange]);

  return (
    <div className="account-block margin_tb_s-5">
      <div className="account-block__row">
        <div className="account-block__type">{type}</div>
        <div className="account-block__balance">Balance: {balance}</div>
      </div>
      <div className="account-block__row">
        <CurrencyLabel
          className="account-block__selector"
          currency={currency}
          onClick={handleSelectorClick}
          showArrow
        />
        <input
          className="account-block__input"
          type="number"
          step="0.01"
          placeholder="0.00"
          max="9999999.99"
          pattern={'^\\d*(\\.\\d{0,2})?$'}
          value={Number(change) === 0 ? '' : change}
          onChange={handleChange}
          onBlur={handleBlur}
        />
      </div>
      <div className="account-block__row">
        <div className="account-block__rate">
          {currency.sign}1 = {otherAccount.currency.sign}{currency.rate}
        </div>
        <div className={`account-block__result ${error ? 'account-block__result_error' : ''}`}>
          { setStatus(error, result) }
        </div>
      </div>
    </div>
  );
}

export default AccountBlock;
