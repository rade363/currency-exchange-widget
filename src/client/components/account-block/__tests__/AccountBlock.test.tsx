import React, { useCallback, useState } from 'react';
import { render, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import validateValue from '../../../utils/validate-value';
import AccountBlock from '../index';
import { Account } from '../../../hooks/useExchange/types';
import { RatesTable } from '../../../utils/calculate-rates/types';
import { DECIMAL_ACCURACY } from '../../../constants';

const sampleRatesTable: RatesTable = {
  usdToGbp: 0.753425,
  usdToEur: 0.885115,
  eurToGbp: 0.851217,
  eurToUsd: 1.129797,
  gbpToEur: 1.174789,
  gbpToUsd: 1.327273,
};
const initialFromAccount: Account = {
  balance: 1137.42,
  change: '0.00',
  result: null,
  error: null,
  currency: {
    name: 'USD',
    sign: '$',
    rate: sampleRatesTable.usdToEur,
  },
};
const initialToAccount: Account = {
  balance: 10,
  change: '0.00',
  result: null,
  error: null,
  currency: {
    name: 'EUR',
    sign: 'â‚¬',
    rate: sampleRatesTable.eurToUsd,
  },
};

function TestWrapperComponent() {
  const [accountFrom, setAccountFrom] = useState<Account>(initialFromAccount);
  const [accountTo, setAccountTo] = useState<Account>(initialToAccount);
  const mockedCallBack = jest.fn();

  const updateAccounts = useCallback((value: string) => {
    const newValue = Number(value);

    setAccountFrom(prevAccountFrom => ({
      ...prevAccountFrom,
      change: value,
      result: newValue === 0 ? null : (prevAccountFrom.balance - newValue).toFixed(DECIMAL_ACCURACY),
      error: validateValue(value, prevAccountFrom.balance),
    }));

    setAccountTo(prevAccountTo => {
      const change = (newValue * accountFrom.currency.rate).toFixed(DECIMAL_ACCURACY);
      return {
        ...prevAccountTo,
        change,
        result: newValue === 0 ? null : (prevAccountTo.balance + Number(change)).toFixed(DECIMAL_ACCURACY),
      };
    });
  }, [accountFrom.currency.rate]);

  return (
    <AccountBlock
      type="From"
      currentAccount={accountFrom}
      otherAccount={accountTo}
      onInputChange={updateAccounts}
      handleSelectorClick={mockedCallBack}
    />
  );
}

const setup = () => {
  const utils = render(<TestWrapperComponent />);
  const input = utils.getByPlaceholderText('0.00') as HTMLInputElement;
  return {
    input,
    ...utils,
  };
};

describe('Account Block', () => {
  it('should be initialized with an empty value (empty string)', () => {
    const { input } = setup();
    expect(input.value).toBe('');
  });

  it('should not accept anything but integers and decimals', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: 'text' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '$' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: ' ' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '42' } });
    expect(input.value).toBe('42');

    fireEvent.change(input, { target: { value: '0' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '42.0567' } });
    expect(input.value).toBe('42.0567');

    fireEvent.change(input, { target: { value: '0' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '..' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '.422' } });
    expect(input.value).toBe('.422');
  });

  it('should display nothing in order to show placeholder in case the value is 0', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: '42' } });
    expect(input.value).toBe('42');

    fireEvent.change(input, { target: { value: '0' } });
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: '0.00' } });
    expect(input.value).toBe('');
  });

  it('should convert integer to decimal on blur event', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: '42' } });
    expect(input.value).toBe('42');

    fireEvent.blur(input);
    expect(input.value).toBe('42.00');
  });

  it('should round the decimal accuracy to the required one', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: '42.12345678' } });
    expect(input.value).toBe('42.12345678');

    fireEvent.blur(input);
    expect(input.value).toBe('42.12');
    expect(input.value.split('.')[1].length).toBe(DECIMAL_ACCURACY);
  });

  it('should not change the value in case the decimal accuracy is valid', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: '42.69' } });
    expect(input.value).toBe('42.69');

    fireEvent.blur(input);
    expect(input.value).toBe('42.69');
    expect(input.value.split('.')[1].length).toBe(DECIMAL_ACCURACY);
  });

  it('should render the account block without input value, result and error', () => {
    const mockedCallBack = jest.fn();

    const block = create(
      <AccountBlock
        type="From"
        currentAccount={initialFromAccount}
        otherAccount={initialToAccount}
        onInputChange={mockedCallBack}
        handleSelectorClick={mockedCallBack}
      />,
    );

    const tree = block.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the account block with result in case of valid input value', () => {
    const mockedCallBack = jest.fn();
    const modifiedAccountFrom: Account = {
      ...initialFromAccount,
      change: '200',
      result: '937.42',
    };

    const block = create(
      <AccountBlock
        type="From"
        currentAccount={modifiedAccountFrom}
        otherAccount={initialToAccount}
        onInputChange={mockedCallBack}
        handleSelectorClick={mockedCallBack}
      />,
    );

    const tree = block.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the account block with error in case of invalid input value', () => {
    const mockedCallBack = jest.fn();
    const modifiedAccountFrom: Account = {
      ...initialFromAccount,
      change: '2000',
      result: '-862.58',
      error: validateValue('2000', 1137.42),
    };

    const block = create(
      <AccountBlock
        type="From"
        currentAccount={modifiedAccountFrom}
        otherAccount={initialToAccount}
        onInputChange={mockedCallBack}
        handleSelectorClick={mockedCallBack}
      />,
    );

    const tree = block.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the account block with maximum input value error', () => {
    const mockedCallBack = jest.fn();
    const modifiedAccountFrom: Account = {
      ...initialFromAccount,
      change: '100000000',
      result: '-99998862,58',
      error: validateValue('100000000', 1137.42),
    };

    const block = create(
      <AccountBlock
        type="From"
        currentAccount={modifiedAccountFrom}
        otherAccount={initialToAccount}
        onInputChange={mockedCallBack}
        handleSelectorClick={mockedCallBack}
      />,
    );

    const tree = block.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
