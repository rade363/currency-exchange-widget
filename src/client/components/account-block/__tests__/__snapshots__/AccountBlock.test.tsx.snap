// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Block should render the account block with error in case of invalid input value 1`] = `
<div
  className="account-block margin_tb_s-5"
>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__type"
    >
      From
    </div>
    <div
      className="account-block__balance"
    >
      Balance: 
      1137.42
    </div>
  </div>
  <div
    className="account-block__row"
  >
    <button
      className="currency-label account-block__selector"
      onClick={[MockFunction]}
      type="button"
    >
      <div
        className="currency-sign currency-label__sign"
      >
        $
      </div>
      <div
        className="currency-label__name"
      >
        USD
      </div>
      <img
        alt="arrow-down"
        className="currency-label__arrow"
        src="test-file-stub"
      />
    </button>
    <input
      className="account-block__input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value="2000"
    />
  </div>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__rate"
    >
      $
      1 = 
      €
      0.89
    </div>
    <div
      className="account-block__result account-block__result_error"
    >
      Insufficient funds: -862.58
    </div>
  </div>
</div>
`;

exports[`Account Block should render the account block with maximum input value error 1`] = `
<div
  className="account-block margin_tb_s-5"
>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__type"
    >
      From
    </div>
    <div
      className="account-block__balance"
    >
      Balance: 
      1137.42
    </div>
  </div>
  <div
    className="account-block__row"
  >
    <button
      className="currency-label account-block__selector"
      onClick={[MockFunction]}
      type="button"
    >
      <div
        className="currency-sign currency-label__sign"
      >
        $
      </div>
      <div
        className="currency-label__name"
      >
        USD
      </div>
      <img
        alt="arrow-down"
        className="currency-label__arrow"
        src="test-file-stub"
      />
    </button>
    <input
      className="account-block__input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value="100000000"
    />
  </div>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__rate"
    >
      $
      1 = 
      €
      0.89
    </div>
    <div
      className="account-block__result account-block__result_error"
    >
      Max value: 8 digits
    </div>
  </div>
</div>
`;

exports[`Account Block should render the account block with result in case of valid input value 1`] = `
<div
  className="account-block margin_tb_s-5"
>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__type"
    >
      From
    </div>
    <div
      className="account-block__balance"
    >
      Balance: 
      1137.42
    </div>
  </div>
  <div
    className="account-block__row"
  >
    <button
      className="currency-label account-block__selector"
      onClick={[MockFunction]}
      type="button"
    >
      <div
        className="currency-sign currency-label__sign"
      >
        $
      </div>
      <div
        className="currency-label__name"
      >
        USD
      </div>
      <img
        alt="arrow-down"
        className="currency-label__arrow"
        src="test-file-stub"
      />
    </button>
    <input
      className="account-block__input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value="200"
    />
  </div>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__rate"
    >
      $
      1 = 
      €
      0.89
    </div>
    <div
      className="account-block__result "
    >
      Result: 937.42
    </div>
  </div>
</div>
`;

exports[`Account Block should render the account block without input value, result and error 1`] = `
<div
  className="account-block margin_tb_s-5"
>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__type"
    >
      From
    </div>
    <div
      className="account-block__balance"
    >
      Balance: 
      1137.42
    </div>
  </div>
  <div
    className="account-block__row"
  >
    <button
      className="currency-label account-block__selector"
      onClick={[MockFunction]}
      type="button"
    >
      <div
        className="currency-sign currency-label__sign"
      >
        $
      </div>
      <div
        className="currency-label__name"
      >
        USD
      </div>
      <img
        alt="arrow-down"
        className="currency-label__arrow"
        src="test-file-stub"
      />
    </button>
    <input
      className="account-block__input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.00"
      type="text"
      value=""
    />
  </div>
  <div
    className="account-block__row"
  >
    <div
      className="account-block__rate"
    >
      $
      1 = 
      €
      0.89
    </div>
    <div
      className="account-block__result "
    >
      
    </div>
  </div>
</div>
`;
